packages/router/src/grpc/view-protocol-server/owned-position-ids.ts

```diff
@@ -0,0 +1,29 @@
+import type { Impl } from '.';
+import { servicesCtx } from '../../ctx';
+import {
+  Position,
+  PositionId,
+} from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/component/dex/v1/dex_pb';
+
+export const ownedPositionIds: Impl['ownedPositionIds'] = async function* (req, ctx) {
+  const services = ctx.values.get(servicesCtx);
+
+  const { indexedDb } = await services.getWalletServices();
+
+  const responses: { positionId: PositionId }[] = [];
+  for await (const positionRecord of indexedDb.iteratePositions()) {
+    const position = Position.fromJson(positionRecord.position);
+
+    if (req.positionState && !req.positionState.equals(position.state)) {
+      continue;
+    }
+    if (req.tradingPair && !req.tradingPair.equals(position.phi?.pair)) {
+      continue;
+    }
+    responses.push({ positionId: PositionId.fromJson(positionRecord.id) });
+  }
+  // direct streaming from indexed-db is not supported because it leads to premature closing of IDB transaction
+  for (const response of responses) {
+    yield response;
+  }
```

Comment:
`ownedPositionIds` was broken for the same reason as #495 
At the same time, I can't do the temporary buffer trick for both `swaps` and `notes` 
https://github.com/penumbra-zone/web/blob/ad3eab64a80f0cdcee3304168c9d46738edfd100/packages/storage/src/indexed-db/index.ts#L207-L216
Because the `POSITIONS` table stores `PositionRecord`, which is not a buf message type. 

Also, I think the buffer solution, although it looks cleaner, results in reading all indexed-db records into memory, just as if we just used `getAll()`

Reply:
Ah, I see. We might as well do a `getAll()`. 

Actually, @turbocrime, can you weigh in here? I'm not sure there is a case to continue doing `streamToGenerator` as it just loads everything into memory all at once anyway.

Reply:
streamToGenerator does not load everything into memory all at once - it streams asynchronously. additionally, streamToGenerator and readable streams are not restricted to protobuf message types.

i wrote idbcursorsource to use a protobuf message type, which i guess was a restrictive choice. but it can be rewritten to support more things, or another simple stream source can be written.

it may also be possible to build indicies for this table that allow more specific queries.

Reply:
if you store jsonified position in idb instead of creating `PositionRecord`, which contains only a jsonified position and a copy of the table key, the pattern is the same as the other functions.

Reply:
you could write a stream like

```ts
  async *getOwnedPositionIds(
    positionState: PositionState | undefined,
    tradingPair: TradingPair | undefined,
  ): AsyncGenerator<PositionId, void> {
    yield* streamToGenerator(
      new ReadableStream({
        start: async cont => {
          let cursor = await this.db.transaction('POSITIONS').store.openCursor();
          while (cursor) {
            const position = Position.fromJson(cursor.value.position);
            if (positionState && !positionState.equals(position.state)) continue;
            if (tradingPair && !tradingPair.equals(position.phi?.pair)) continue;
            cont.enqueue(PositionId.fromJson(cursor.value.id));
            cursor = await cursor.continue();
          }
          cont.close();
        },
      }),
    );
  }
  ```

Reply:
Okay, `streamToGenerator` does not cause all records to be loaded into memory. 
I implemented getOwnedPositionIds() following the example from @turbocrime , and modified the tests.